judge/views/user.py

```diff
@@ -507,3 +512,106 @@ def post(self, request, *args, **kwargs):
+            'protocol': 'https' if self.request.is_secure() else 'http',
+            'site_admin_email': settings.SITE_ADMIN_EMAIL,
+            'expiry_minutes': settings.DMOJ_EMAIL_CHANGE_EXPIRY_MINUTES,
+            'user': self.request.user,
+            'activation_key': activation_key,
+            'new_email': new_email,
+        }
+        # When from_email is none, we use the DEFAULT_FROM_EMAIL setting
+        send_mail(
+            self.notify_subject_template_name, self.notify_email_template_name, context, from_email=None,
+            to_email=self.request.user.email, html_email_template_name=self.notify_html_email_template_name,
+        )
+        send_mail(
+            self.activate_subject_template_name, self.activate_email_template_name, context, from_email=None,
+            to_email=new_email, html_email_template_name=self.activate_html_email_template_name,
+        )
+
+        return generic_message(
+            self.request,
+            _('Email change requested'),
+            _('Please click on the link sent to the new email address.'),
+        )
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs['user'] = self.request.user
+        return kwargs
+
+    def post(self, request, *args, **kwargs):
+        key = f'emailchange!{request.META["REMOTE_ADDR"]}'
+        cache.add(key, 0, timeout=settings.DMOJ_EMAIL_CHANGE_LIMIT_WINDOW)
+        if cache.incr(key) > settings.DMOJ_EMAIL_CHANGE_LIMIT_COUNT:
+            return HttpResponse(_('You have sent too many email change requests. Please try again later.'),
+                                content_type='text/plain', status=429)
+        return super().post(request, *args, **kwargs)
+
+
+class EmailChangeActivateView(LoginRequiredMixin, View):
+    class EmailChangeFailedError(Exception):
+        pass
+
+    def update_user_email(self, request, activation_key):
+        signer = signing.TimestampSigner()
+        try:
+            data = signer.unsign_object(
+                base64.urlsafe_b64decode(activation_key.encode()).decode(),
+                max_age=settings.DMOJ_EMAIL_CHANGE_EXPIRY_MINUTES * 60,
+            )
+        except (binascii.Error, signing.BadSignature):
+            raise self.EmailChangeFailedError(_('Invalid activation key. Please try again.'))
+        except signing.SignatureExpired:
+            raise self.EmailChangeFailedError(_('This request has expired. Please try again.'))
+        if data['id'] != request.user.id:
+            raise self.EmailChangeFailedError(
+                _('Please try again from the account this email change was originally requested from.'),
+            )
+        with revisions.create_revision(atomic=True):
+            if User.objects.filter(email=data['email']).exists():
+                raise self.EmailChangeFailedError(
+                    _('The email originally requested is in use. Please try again with a new email.'),
```

Comment:
```suggestion
                    _('The email you originally requested has since been registered by another user. Please try again with a new email.'),
```